pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,261 @@
+
+package coscheduling
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/api/v1/pod"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+)
+
+// Coscheduling is a plugin that implements the mechanism of gang scheduling.
+type Coscheduling struct {
+	FrameworkHandle framework.FrameworkHandle
+	PodLister       corelisters.PodLister
+	PodGroupInfos   sync.Map // Key is the name of Namespace/PodGroup.
+	SeqNum          uint64
+}
+
+// TODO implement a timeout based gc for the PodGroupInfos map
+type PodGroupInfo struct {
+	// Timestamp stores the timestamp of the initialization time of PodGroup.
+	Timestamp    time.Time
+	SeqNum       uint64
+	MinAvailable int
+	Name         string
+}
+
+var _ framework.QueueSortPlugin = &Coscheduling{}
+var _ framework.PreFilterPlugin = &Coscheduling{}
+var _ framework.PermitPlugin = &Coscheduling{}
+var _ framework.UnreservePlugin = &Coscheduling{}
+
+const (
+	// Name is the name of the plugin used in Registry and configurations.
+	Name                 = "coscheduling"
+	PodGroupName         = "pod-group.scheduling.sigs.k8s.io/name"
+	PodGroupMinAvailable = "pod-group.scheduling.sigs.k8s.io/min-available"
+	PermitWaitingTime    = 1 * time.Second
+)
+
+// Name returns name of the plugin. It is used in logs, etc.
+func (cs *Coscheduling) Name() string {
+	return Name
+}
+
+// New initializes a new plugin and returns it.
+func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin, error) {
+	podLister := handle.SharedInformerFactory().Core().V1().Pods().Lister()
+	return &Coscheduling{FrameworkHandle: handle,
+		PodLister: podLister,
+	}, nil
+}
+
+// Less are used to sort pods in the scheduling queue.
+// 1. compare the priority of pod
+// 2. compare the timestamp of the initialization time of PodGroup
+// 3. compare the SeqNum
+func (cs *Coscheduling) Less(podInfo1 *framework.PodInfo, podInfo2 *framework.PodInfo) bool {
+	pod1 := podInfo1.Pod
+	pod2 := podInfo2.Pod
+	priority1 := pod.GetPodPriority(pod1)
+	priority2 := pod.GetPodPriority(pod2)
+
+	if priority1 != priority2 {
+		return priority1 > priority2
+	}
+
+	pgInfo1 := cs.getPodGroupInfo(podInfo1)
+	pgInfo2 := cs.getPodGroupInfo(podInfo2)
+	time1 := pgInfo1.Timestamp
+	time2 := pgInfo2.Timestamp
+
+	if pgInfo1.SeqNum > 0 && pgInfo2.SeqNum > 0 && time1.Equal(time2) {
+		return pgInfo1.SeqNum < pgInfo2.SeqNum
+	}
+	return time1.Before(time2)
+}
+
+func (cs *Coscheduling) getPodGroupInfo(p *framework.PodInfo) *PodGroupInfo {
+	podGroupName, min := GetPodGroupLabels(p.Pod)
+	if podGroupName != "" {
+		key := fmt.Sprintf("%v/%v", p.Pod.Namespace, podGroupName)
+		pgInfo, ok := cs.PodGroupInfos.Load(key)
+		if !ok {
+			pgInfo = &PodGroupInfo{
+				Name:         podGroupName,
+				Timestamp:    p.InitialAttemptTimestamp,
```

Comment:
Given that the resolution of Timestamp is very good, I don't think we need SeqNum. We can just do any arbitrary comparison, like alphabetical order of the pod group name.

Reply:
At the beginning, I want to use some other comparison methods like alphabetic string comparison. But later, I think the auto increment key is easier to understand.

Reply:
but that's already conveyed in the Timestamp.